Add documentation about nodes with generator functions (#2302)
kedro-org/kedro: docs/source/nodes_and_pipelines/nodes.md

```{note}
You can also call a node as a regular Python function: `adder_node(dict(a=2, b=3))`. This will call `adder_node.run(dict(a=2, b=3))` behind the scenes.
```

## How to use generator functions in a node

> Generator functions are useful for providing lazy iterators over big number of items without storing all of them in memory at once. We can wrap generator functions in a Kedro pipeline, the Kedro runner will make sure to save the result by calling the dataset's `save` method separately for each chunk.

Reviewer: ```suggestion
Generator functions were introduced with [PEP 255](https://www.python.org/dev/peps/pep-0255). They are are a special kind of function that returns a lazy iterators but do not store their entire contents in memory all at once.

We can wrap generator functions in a Kedro pipeline. The Kedro runner takes care of saving the result by calling the dataset's `save` method separately for each chunk.
```

> The following tutorial will show you how to utilise `pandas chunksize` generator to process large datasets, we will demonstrate this using the [`pandas-iris`](https://github.com/kedro-org/kedro-starters/tree/main/pandas-iris) [starter](https://kedro.readthedocs.io/en/stable/kedro_project_setup/starters.html):

Reviewer: ```suggestion
The following tutorial shows how to use a `pandas chunksize` generator to process large datasets within the [`pandas-iris`](https://github.com/kedro-org/kedro-starters/tree/main/pandas-iris) [starter](https://kedro.readthedocs.io/en/stable/kedro_project_setup/starters.html):
```

Reviewer: @SajidAlamQB When you link to other pages of docs, you need to use a relative link rather than hard code the URL of the docs. I've corrected the one above but can you check your others please?

Author: Will do, thank you! 🙏

Reviewer: I would then add a section which explains how to create the project. So I'd say something about the fact we assume you have already set up Kedro--something like this:

If you have not yet set up Kedro, do so by [following the guidelines to install Kedro](../get_started/install.md).)

Then, link to [the page that explains how to create a new project](../get_started/new_project.md#create-the-example-project) to show how to make the project.

Author: I will link to how to use `Kedro starters` as that should cover how to create the `pandas-iris` project properly.

> Create a [custom dataset](https://kedro.readthedocs.io/en/stable/extend_kedro/custom_datasets.html) called `ChunkWiseCSVDataSet` in `src/YOUR_PROJECT_NAME/extras/datasets/chunkwise_datasets.py` for your `pandas-iris` project. This dataset is a simplified version of the `pandas.CSVDataSet` where the main change is to the `_save` method which should save the data in append-or-create mode, `a+`. 

Reviewer: Please fix this URL

            filepath=PurePosixPath(path),
            version=version,
            exists_function=self._fs.exists,
            glob_function=self._fs.glob,
        )

        # Handle default load and save arguments
        self._load_args = deepcopy(self.DEFAULT_LOAD_ARGS)
        if load_args is not None:
            self._load_args.update(load_args)
        self._save_args = deepcopy(self.DEFAULT_SAVE_ARGS)
        if save_args is not None:
            self._save_args.update(save_args)

    def _describe(self) -> Dict[str, Any]:
        return {
            "filepath": self._filepath,
            "protocol": self._load_args,
            "save_args": self._save_args,
            "version": self._version,
        }

    def _load(self) -> pd.DataFrame:
        load_path = str(self._get_load_path())
        return pd.read_csv(load_path, **self._load_args)

    def _save(self, data: pd.DataFrame) -> None:
        save_path = get_filepath_str(self._get_save_path(), self._protocol)

        buf = BytesIO()
        data.to_csv(path_or_buf=buf, **self._save_args)

        with self._fs.open(save_path, mode="a+") as fs_file:
            fs_file.write(buf.getvalue())

> ```

Reviewer: That's quite a long chunk of code (above) and I think you need to put it inside a "Click here to expand" block.

> Modify `example_iris_data` in the `catalog.yml` by changing `type` to the custom dataset we created above and adding `chunksize: 100` to `load_args` which will return an iterable object. The `chunksize` parameter itself refers to the number of rows in each chunk.

Reviewer: ```suggestion
Modify `example_iris_data` in `catalog.yml` by changing `type` to the custom dataset you created above. Add `chunksize: 100` to `load_args` which will return an iterable object. The `chunksize` parameter refers to the number of rows in each chunk.
```

```yaml
example_iris_data:
  type: YOUR_PROJECT_NAME.extras.datasets.chunkwise_dataset.ChunkWiseCSVDataSet
  filepath: data/01_raw/iris.csv
  load_args:
    chunksize: 100
```

> In `nodes.py` we will repurpose the [`split_data`](https://github.com/kedro-org/kedro-starters/blob/dacdd56f1c1afde00a03a1e342fc0f44e1567a1e/pandas-iris/%7B%7B%20cookiecutter.repo_name%20%7D%7D/src/%7B%7B%20cookiecutter.python_package%20%7D%7D/nodes.py#L13) function to process chunk-wise data:

Reviewer: ```suggestion
Next, in `nodes.py` we repurpose the [`split_data`](https://github.com/kedro-org/kedro-starters/blob/dacdd56f1c1afde00a03a1e342fc0f44e1567a1e/pandas-iris/%7B%7B%20cookiecutter.repo_name%20%7D%7D/src/%7B%7B%20cookiecutter.python_package%20%7D%7D/nodes.py#L13) function to process chunk-wise data:
```

```python
def split_data(
    data: pd.DataFrame, parameters: Dict[str, Any]
) -> Tuple[pd.DataFrame, pd.DataFrame, pd.Series, pd.Series]:
    """Splits data into features and target training and test sets.

    Args:
        data: Data containing features and target.
        parameters: Parameters defined in parameters.yml.
    Returns:
        Split data.
    """
    # Loop through data in chunks building up the training and test sets
    for chunk in data: # Iterate over the chunks from data
        full_data = pd.concat([chunk]) # Converts the TextFileReader object into list of DataFrames
        data_train = full_data.sample(
            frac=parameters["train_fraction"], random_state=parameters["random_state"]
        )
        data_test = full_data.drop(data_train.index)

        X_train = data_train.drop(columns=parameters["target_column"])
        X_test = data_test.drop(columns=parameters["target_column"])
        y_train = data_train[parameters["target_column"]]
        y_test = data_test[parameters["target_column"]]
        yield X_train, X_test, y_train, y_test
```

> We can now `kedro run` in the terminal where the output will be saving `X_train`, `X_test`, `y_train`, `y_test` in chunks:

Reviewer: ```suggestion
We can now `kedro run` in the terminal. The output shows `X_train`, `X_test`, `y_train`, `y_test` saved in chunks:
```

```{note}
You can also call a node as a regular Python function: `adder_node(dict(a=2, b=3))`. This will call `adder_node.run(dict(a=2, b=3))` behind the scenes.
```

## How to use generator functions in a node

> [Generator functions](https://learnpython.org/en/Generators) were introduced with [PEP 255](https://www.python.org/dev/peps/pep-0255). They are a special kind of function that returns a lazy iterators but do not store their entire contents in memory all at once.

Reviewer: ```suggestion
[Generator functions](https://learnpython.org/en/Generators) were introduced with [PEP 255](https://www.python.org/dev/peps/pep-0255). They are a special kind of function that returns lazy iterators but do not store their entire contents in memory all at once.
```

> The following code uses a `pandas chunksize` generator to process large datasets within the [`pandas-iris` starter](../kedro_project_setup/starters.md). Going forward we will assume you have already set up a Kedro `pandas-iris` starter project. If you have not yet, do so by following the [Kedro starters guidelines](../kedro_project_setup/starters.md).

Reviewer: ```suggestion
The following code uses a `pandas chunksize` generator to process large datasets within the [`pandas-iris` starter](../kedro_project_setup/starters.md). We assume you have already set up a Kedro `pandas-iris` starter project. If you have not yet, do so by following the [get started guide](../[kedro_project_setup/starters.md](../get_started/new_project.md#create-the-example-project).
```

```{note}
You can also call a node as a regular Python function: `adder_node(dict(a=2, b=3))`. This will call `adder_node.run(dict(a=2, b=3))` behind the scenes.
```

## How to use generator functions in a node

> Generator functions are useful for providing lazy iterators to search for items without storing in memory. We can wrap generator functions in a Kedro pipeline and save the result by calling the dataset's `save` method separately for each chunk.

Reviewer 2: ```suggestion
Generator functions are useful for providing lazy iterators over big number of items without storing all of them in memory at once. We can wrap generator functions in a Kedro pipeline and the Kedro runner will make sure to save the result by calling the dataset's `save` method separately for each chunk.
```

The following example will show you how to utilise `pandas chunksize` generator to process large datasets:

> We need to create a [custom dataset](https://kedro.readthedocs.io/en/stable/extend_kedro/custom_datasets.html) with `_load` having `chunksize` parameter for `pd.read_csv()` and `_save` should save the data in append-or-create mode, `a+`.

Reviewer 2: Do we really need to change the `_load` method? Can't we just provide it in `load_args` in the catalog?

Could we show the code for the full custom dataset here? I suspect it just needs to be wrapped in a class like this:
```
class ChunkWiseCSVDataSet(pandas.CSVDataSet):
    ....
```

It would be nice if the user can follow this and test the functionality without too much guessing what needs to go where.

> We need to create a [custom dataset](https://kedro.readthedocs.io/en/stable/extend_kedro/custom_datasets.html) with `_load` having `chunksize` parameter for `pd.read_csv()` and `_save` should save the data in append-or-create mode, `a+`.

Author: Yes, it's a custom `pandas.CSVDataSet`.

```python
def _load(self) -> pd.DataFrame:
    load_path = str(self._get_load_path())
    return pd.read_csv(load_path, **self._load_args, chunksize=50)

def _save(self, data: pd.DataFrame) -> None:
    save_path = get_filepath_str(self._get_save_path(), self._protocol)

    buf = BytesIO()
    data.to_csv(path_or_buf=buf, **self._save_args)

    with self._fs.open(save_path, mode="a+") as fs_file:
        fs_file.write(buf.getvalue())
```

> Using the `pandas-iris` starter as an example we will repurpose `split_dataset` function to process chunk-wise data:

Reviewer 3: ```suggestion
Using the `pandas-iris` starter as an example we will repurpose the `split_dataset` function in `nodes.py` to process chunk-wise data:
```

```python
def split_data(
    data: pd.DataFrame, parameters: Dict[str, Any]
) -> Tuple[pd.DataFrame, pd.DataFrame, pd.Series, pd.Series]:
    """Splits data into features and target training and test sets.

    Args:
        data: Data containing features and target.
        parameters: Parameters defined in parameters.yml.
    Returns:
        Split data.
    """
    for chunk in data:
        full_data = pd.concat([chunk])
        data_train = full_data.sample(
            frac=parameters["train_fraction"], random_state=parameters["random_state"]
        )
        data_test = full_data.drop(data_train.index)

        X_train = data_train.drop(columns=parameters["target_column"])
        X_test = data_test.drop(columns=parameters["target_column"])
        y_train = data_train[parameters["target_column"]]
        y_test = data_test[parameters["target_column"]]

>         yield X_train, X_test, y_train, y_test

Reviewer 3: It would be helpful to add some comments in here to explain what's happening. That way users can also more easily figure out what steps they need to add to their own code.

Reviewer 2: Also it would be nice to show the contrast from the original, maybe a git diff?

Author: The git diff didn't look that great IMO, so I've left some comments explaining the additions and changes. It isn't too different from the original version which is linked.

```

> When we now do `kedro run` the output should now be saving `X_train`, `X_test`, `y_train`, `y_test` in chunks:

Reviewer 3: ```suggestion
When we now do `kedro run` the output will be saving `X_train`, `X_test`, `y_train`, `y_test` in chunks:
```
